File: d15/d15p2.js
const TEST_CASES = [
    // { input: '0,3,6', expected: 175594 },
]


function solve(input) {
    const starting_numbers = input.trim().split(',').map(Number);
    const spoken = new Map();
    starting_numbers.slice(0,-1).forEach((n,i) => spoken.set(n, i+1));
    let last_spoken = starting_numbers[starting_numbers.length -1];

    for(let turn=starting_numbers.length; turn < 30000000; turn+=1) {
        if (spoken.has(last_spoken)) {
            next = turn - spoken.get(last_spoken);
        } else {
            next = 0
        }
        spoken.set(last_spoken, turn);
        last_spoken = next;
    }
    return last_spoken;
}

function main() {
    TEST_CASES.forEach(({input, expected}) => {
        console.assert(expected === solve(input));
        console.log('OK');
    });
    console.log(solve('15,5,1,4,7,0'))
}

main();

File: d15/d15p2.py
import time

from utils.test_case import TestCase
from d15_input import INPUT

TEST_CASES = [
    TestCase("0,3,6", 175594),
    # TestCase("1,3,2", 1),
    # TestCase("2,1,3", 10),
    # TestCase("1,2,3", 27),
    # TestCase("2,3,1", 78),
    # TestCase("3,2,1", 438),
    # TestCase("3,1,2", 1836),
]


def solve(input):
    starting_numbers = [int(n) for n in input.strip().split(',')]
    spoken = {number: i+1 for i, number in enumerate(starting_numbers[:-1])}
    last_spoken = starting_numbers[-1]

    for turn in range(len(starting_numbers), 30000000):
        if last_spoken in spoken:
            next = turn - spoken[last_spoken]
        else:
            next = 0

        spoken[last_spoken] = turn
        # print(f'turn {turn}: spoken: {last_spoken}')
        last_spoken = next

    return last_spoken


if __name__ == '__main__':
    for case in TEST_CASES:
        result = solve(case.case)
        case.check(result)

    t0 = time.time()
    print(solve(INPUT))
    t1 = time.time()
    print(f"{(t1 - t0) * 1000:0.1f} ms")
